### PR TITLE
Cascade remove

### PR DESCRIPTION
**Pull request by rriclet**

### CODE REVIEW

File: migration/src/main/scala/org/thp/thehive/migration/th3/Conversion.scala [hunk 1]
```diff
@@ -151,12 +151,11 @@ trait Conversion {
       metaData <- json.validate[MetaData]
       message  <- (json \ "message").validate[String]
       date     <- (json \ "startDate").validate[Date]
-      deleted = (json \ "status").asOpt[String].contains("Deleted")
       attachment =
```
[the remark this answers is not in the document]
Comment (rriclet) [reply migration/src/main/scala/org/thp/thehive/migration/th3/Conversion.scala:154]: But we can't create a log with a `deleted` property ? (line 158)

File: thehive/app/org/thp/thehive/services/ShareSrv.scala [hunk 2]
```diff
@@ -90,19 +91,44 @@ class ShareSrv @Inject() (
...
 
   def remove(shareId: EntityIdOrName)(implicit graph: Graph, authContext: AuthContext): Try[Unit] =
+    unshareCase(shareId).flatMap(_ => Try(get(shareId).remove()))
+
+  def cascadeRemove(shareId: EntityIdOrName)(implicit graph: Graph, authContext: AuthContext): Try[Unit] = {
+    for {
+      organisation <- organisationSrv.getOrFail(authContext.organisation)
+      share        <- getOrFail(shareId)
+
+      tasks          <- Try(get(share).tasks)
```
Comment (To-om) [inline thehive/app/org/thp/thehive/services/ShareSrv.scala:101]: `Try()` are pointlesss. If the method doesn't return a Try use: `tasks = get(share).tasks`

File: thehive/app/org/thp/thehive/services/ShareSrv.scala [hunk 3]
```diff
@@ -90,19 +91,44 @@ class ShareSrv @Inject() (
...
+      _              <- tasksToUnshare.toTry(unshareTask(_, organisation))
+      _              <- tasksToDelete.toTry(taskSrv.cascadeRemove(_, share))
+
+      obs          <- Try(get(share).observables)
+      obsToUnshare <- Try(obs.clone().filter(_.inE[ShareObservable].count.is(P.gt(1))).toSeq)
+      obsToDelete  <- Try(obs.filter(_.inE[ShareObservable].count.is(P.eq(1))).toSeq)
+      _            <- obsToUnshare.toTry(unshareObservable(_, organisation))
+      _            <- obsToDelete.toTry(observableSrv.cascadeRemove(_, share))
+
+      cazeDelete   <- Try(get(share).`case`.filter(_.inE[ShareCase].count.is(P.eq(1))).toSeq)
```
Comment (To-om) [inline thehive/app/org/thp/thehive/services/ShareSrv.scala:113]: cazeDelete is empty or contains only one element => use Option instead of Seq for clarity

**Review by To-om — CHANGES_REQUESTED** [review]

File: thehive/app/org/thp/thehive/services/ObservableSrv.scala [hunk 4]
```diff
@@ -174,9 +174,23 @@ class ObservableSrv @Inject() (
...
           .map(_ => ())
-      case Some(alert) => alertSrv.removeObservable(alert, observable)
+      case Some(alert) => for {
+        _ <- Try(get(observable).remove())
+      } yield auditSrv.observableInAlert.delete(observable, Some(alert))
     }
 
+  // Same as remove but with no Audit creation
+  def cascadeRemove(observable: Observable with Entity, share: Share with Entity)(implicit graph: Graph, authContext: AuthContext): Try[Unit] = {
+    val alert = get(observable).alert.headOption
```
[the remark this answers is not in the document]
Comment (rriclet) [reply thehive/app/org/thp/thehive/services/ObservableSrv.scala:184]: Indeed, `cascadeRemove` is only called from `ShareSrv`.
Deletion for an Alert observable is handled is the function above : `remove(obs: Observable with Entity)`
Should we merge the two functions ?

File: thehive/app/org/thp/thehive/services/ShareSrv.scala (hunk 2 above)
Comment (rriclet) [reply thehive/app/org/thp/thehive/services/ShareSrv.scala:101]: Not wrapping in `Try` result in an error : Traversal[Nothing, Nothing, Nothing] is found but Try[Unit] is expected

**Pull request merged**